File: backend/hip3_dexes.py
"""HIP-3 venue identity for the API.

Protocol code uses `{dex}:{COIN}` (e.g. `xyz:SNDK`, `io:ANTH`). Enabling a dex
here means we fetch that clearinghouse / meta — catalog tickers are still
allowlisted in `ASSET_METADATA` (with optional `dex`, default `xyz`).
"""

from __future__ import annotations

import os
import re

DEFAULT_HIP3_DEXES = ("xyz", "io")
_DEX_NAME_RE = re.compile(r"^[a-z]{2,4}$")


def parse_enabled_hip3_dexes(raw: str | None) -> list[str]:
    fallback = list(DEFAULT_HIP3_DEXES)
    if not raw or not str(raw).strip():
        return fallback
    out: list[str] = []
    seen: set[str] = set()
    for part in str(raw).split(","):
        d = part.strip().lower()
        if not _DEX_NAME_RE.match(d) or d in seen:
            continue
        seen.add(d)
        out.append(d)
    return out or fallback


def enabled_hip3_dexes() -> list[str]:
    return parse_enabled_hip3_dexes(os.environ.get("HIP3_ENABLED_DEXES"))


def is_hip3_dex_name(name: str | None) -> bool:
    n = (name or "").strip()
    if not n:
        return False
    lower = n.lower()
    if lower == "trade.xyz":
        return True
    if lower in {d.lower() for d in enabled_hip3_dexes()}:
        return True
    return n == lower and bool(_DEX_NAME_RE.match(n))


def split_hip3_coin(value: str, fallback_dex: str = "xyz") -> tuple[str, str]:
    """Return `(dex, base)` for HL `xyz:SNDK` or alert storage `SNDK:xyz`."""
    raw = (value or "").strip()
    fallback = (fallback_dex or "xyz").lower()
    if ":" not in raw:
        return fallback, raw
    left, right = raw.split(":", 1)
    if is_hip3_dex_name(left):
        return left.lower(), right
    if is_hip3_dex_name(right):
        return right.lower(), left
    return fallback, right or left


def hip3_display_symbol(symbol: str) -> str:
    raw = (symbol or "").strip()
    if not raw or ":" not in raw:
        return raw
    _dex, base = split_hip3_coin(raw)
    return base or raw

File: frontend/src/lib/hip3Dexes.ts
/**
 * HIP-3 venue identity. Protocol code must use `{dex}:{COIN}` (e.g. `xyz:SNDK`,
 * `io:ANTH`) — never a bare ticker as a market id.
 *
 * Catalog tickers are still allowlisted in `ASSET_METADATA`. Enabling a dex
 * here only means we subscribe/fund/fetch that clearinghouse. Forks set
 * `EXPO_PUBLIC_HIP3_ENABLED_DEXES` (comma-separated, 2–4 letter names).
 */

export const DEFAULT_HIP3_DEXES = ['xyz', 'io'] as const;

const DEX_NAME_RE = /^[a-z]{2,4}$/;

export function parseEnabledHip3Dexes(raw?: string | null): string[] {
  const fallback = [...DEFAULT_HIP3_DEXES];
  if (!raw || !String(raw).trim()) return fallback;
  const out: string[] = [];
  const seen = new Set<string>();
  for (const part of String(raw).split(',')) {
    const d = part.trim().toLowerCase();
    if (!DEX_NAME_RE.test(d) || seen.has(d)) continue;
    seen.add(d);
    out.push(d);
  }
  return out.length ? out : fallback;
}

export function enabledHip3Dexes(): string[] {
  return parseEnabledHip3Dexes(process.env.EXPO_PUBLIC_HIP3_ENABLED_DEXES);
}

/** True for an HL perp-dex name (enabled list, or live lowercase 2–4 letter prefix). */
export function isHip3DexName(name: string | null | undefined): boolean {
  const n = String(name ?? '').trim();
  if (!n) return false;
  const lower = n.toLowerCase();
  if (lower === 'trade.xyz') return true;
  if (enabledHip3Dexes().includes(lower)) return true;
  // HL coin form is `xyz:SNDK` / `io:ANTH` (lowercase dex). Do not treat
  // uppercase tickers (`CL`, `GOLD`) as dex names.
  return n === lower && DEX_NAME_RE.test(n);
}

export function splitHip3Symbol(
  value: string,
  fallbackDex?: string | null,
): { dex: string; base: string } {
  const raw = String(value ?? '').trim();
  const fallback = (fallbackDex || 'xyz').toLowerCase();
  if (!raw.includes(':')) {
    return { dex: fallback, base: raw };
  }
  const [left, ...rest] = raw.split(':');
  const right = rest.join(':');
  if (isHip3DexName(left)) {
    return { dex: left.toLowerCase(), base: right };
  }
  if (isHip3DexName(right)) {
    return { dex: right.toLowerCase(), base: left };
  }
  return { dex: fallback, base: right || left };
}

/** UI label: `xyz:SNDK` / `SNDK:xyz` / `io:ANTH` → `SNDK` / `ANTH`. Never empty. */
export function hip3DisplaySymbol(symbol: string): string {
  const raw = String(symbol ?? '').trim();
  if (!raw) return raw;
  if (!raw.includes(':')) return raw;
  const { base } = splitHip3Symbol(raw);
  return base || raw;
}

export function normalizeHip3Coin(coin: string, dex?: string | null): string {
  const parsed = splitHip3Symbol(coin, dex);
  return `${parsed.dex}:${parsed.base}`;
}
